### PR TITLE
Cli setup

### PR DESCRIPTION
**Pull request by adamshire123**
### What does this PR do?

app gets config settings from environment variables or defaults.
adds cli commands and options

### Includes new or updated dependencies?

YES

### What are the relevant tickets?
https://mitlibraries.atlassian.net/browse/IN-732

### Developer

- [ ] All new ENV is documented in README (or there is none)
- [ ] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer

- [ ] The commit message is clear and follows our guidelines
      (not just this pull request message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The documentation has been updated or is unnecessary
- [ ] The changes have been verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

**Comment by adamshire123** [comment]
I think something in cli.py is breaking pylama, but I'm not sure what.

**Comment by ehanson8** [comment]
It appears to be related to an incompatibility between `pylama` and `pydocstyle` due to some recent changes ([link](https://github.com/klen/pylama/issues/232)). We can remove `pydocstyle` from the `pylama` linters in `setup.cfg` as a quick fix but there's other linting error hiding behind that. And @hakbailey may have a better solution but it's annoying either way when we have to pin a version # or remember to add a linter back in.

Slightly unrelated, both Adam and I found out last week that running `isort` through `pylama` gives a very unhelpful message (`Incorrectly sorted imports. [isort]`) so we might consider restoring it as a separate linter in the template repo `Makefile` to get the useful messages back

File: setup.cfg [hunk 1]
```diff
@@ -8,7 +8,7 @@ ignore_missing_imports = True
 
 [pylama]
 ignore = C0114,C0116,D100,D103,W0012
-linters = eradicate,isort,mccabe,pycodestyle,pydocstyle,pyflakes,pylint
+linters = eradicate,isort,mccabe,pycodestyle,pyflakes,pylint
```
Comment (hakbailey) [inline setup.cfg:11]: If we're going to remove pydocstyle from the pylama checks, then we should add it as its own dev dependency and Makefile command for linting.

File: sapinvoices/cli.py [hunk 2]
```diff
@@ -1,28 +1,81 @@
...
+    help="USE WITH CAUTION. If '--real-run' flag is passed, files will be emailed "
+    "to stakeholders and, if the '--final-run' flag is also passed, invoices will be "
+    "sent to SAP and marked as paid in Alma. If this flag is not passed, this command "
+    "defaults to a dry run, in which files will not be emailed or sent to SAP, instead "
+    "their contents will be logged for review, and invoices will also not be marked as "
+    "paid in Alma.",
+)
+@click.pass_context
+def sap_invoices(
+    ctx: click.Context, final_run: Optional[str], real_run: Optional[str]
```
Comment (hakbailey) [inline sapinvoices/cli.py]: The types for `final_run` and `real_run` should be `bool`.

File: sapinvoices/config.py [hunk 3]
```diff
@@ -3,21 +3,36 @@
 
 import sentry_sdk
 
+CONFIG = [
```
Comment (hakbailey) [inline sapinvoices/config.py]: I would add `WORKSPACE` and `SSM_PATH` to this (and probably remove `LOG_LEVEL` and `SENTRY_DSN`). Then the `load_config_values()` function can simply use `os.environ` to populate all the fields.

Comment (hakbailey) [reply sapinvoices/config.py]: We also need to add the `ssm_safety_check` functionality somewhere in the config loading process.

File: README.md [hunk 4]
```diff
@@ -13,14 +13,15 @@ sets trasmitted invoice to paid in Alma
 
 ## Required ENV
 
-- `SENTRY_DSN` = If set to a valid Sentry DSN, enables Sentry exception monitoring. This is not needed for local development.
-- `WORKSPACE` = Set to `dev` for local development, this will be set to `stage` and `prod` in those environments by Terraform.
+
 - `ALMA_API_URL` = Base URL for making Alma API calls
 - `ALMA_API_READ_WRITE_KEY` = API key for making Alma API calls
+- `LOG_LEVEL`
```
Comment (hakbailey) [inline README.md]: This should note that `LOG_LEVEL` is optional and will default to `INFO` if not set.

File: sapinvoices/cli.py [hunk 5]
```diff
@@ -1,28 +1,81 @@
...
+    Retrieves "Waiting to be sent" invoices from Alma, extracts and formats data
+    needed for submission to SAP for payment. If not a final run, creates and sends
+    formatted review reports to Acquisitions staff. If a final run, creates and sends
+    formatted cover sheets and summary reports to Acquisitions staff, submits data and
+    control files to SAP, and marks invoices as paid in Alma after submission to SAP.
+    """
+    config_values = load_config_values()
     root_logger = logging.getLogger()
-    logger.info(configure_logger(root_logger, verbose))
+    log_level = config_values["LOG_LEVEL"] or "INFO"
```
Comment (hakbailey) [inline sapinvoices/cli.py]: `or "INFO"` not needed here since it defaults to that in the `configure_logger()` function

File: Pipfile [hunk 6]
```diff
@@ -20,4 +21,4 @@ pytest = "*"
 python_version = "3.11"
 
 [scripts]
-sapinvoices = "python -c \"from sapinvoices.cli import main; main()\""
+sapinvoices = "python -c \"from sapinvoices.cli import cli; cli()\""
```
Comment (hakbailey) [inline Pipfile]: You might consider changing this command to `sap` and the subcommands in the cli main group to something like `create-sandbox-data` and `process-invoices`. Running the full command right now as `pipenv run sapinvoices sap-invoices` feels a little odd.

File: README.md [hunk 7]
```diff
@@ -13,14 +13,15 @@ sets trasmitted invoice to paid in Alma
...
 - `SAP_DROPBOX_CONNECTION` = JSON formatted connection information for accessing SAP dropbox
 - `SAP_REPLY_TO_EMAIL`
```
Comment (hakbailey) [inline README.md]: I would go ahead and write descriptions for each of these.

**Review by hakbailey — CHANGES_REQUESTED** [review]
See inline comments for changes and suggestions.

File: setup.cfg (hunk 1 above)
Comment (ehanson8) [reply setup.cfg:11]: Good call, that is a better solution

File: sapinvoices/cli.py (hunk 5 above)
Comment (adamshire123) [reply sapinvoices/cli.py]: If WORKSPACE=dev and LOG_LEVEL is not set in .env, excluding `or "INFO"` is a key error

File: README.md (hunk 7 above)
Comment (adamshire123) [reply README.md]: what is SES_SEND_FROM_EMAIL?

File: sapinvoices/cli.py (hunk 5 above)
Comment (adamshire123) [reply sapinvoices/cli.py]: I think my intention was to pass this in as an option, so `or "INFO"` is just a place holder for that, though I probably need to reverse which side of the `or` the option is on once I'm accepting the option.  Is there a way to have an option apply to the group?

Comment (ehanson8) [reply sapinvoices/cli.py]: I used `logger.info(configure_logger(root_logger, os.getenv("LOG_LEVEL", "None")))` in an update to my PR which will solve this

File: README.md (hunk 7 above)
Comment (ehanson8) [reply README.md]: The email address that SES sends from, e.g. `noreply@libraries.mit.edu`

File: sapinvoices/config.py (hunk 3 above)
Comment (adamshire123) [reply sapinvoices/config.py:6]: SSM_PATH I left out by mistake. can you tell me more about why you'd add WORKSPACE to the CONFIG settings and not just get it from env? I think we only need WORKSPACE for config.py which is why I left it out of the list of CONFIG settings.

File: tests/conftest.py [hunk 8]
```diff
@@ -9,12 +9,12 @@ def test_env():
...
         "ALMA_API_URL": "test",
         "ALMA_API_READ_WRITE_KEY": "test",
-        "LOG_LEVEL": "test",
+        "LOG_LEVEL": "DEBUG",
         "SAP_DROPBOX_CONNECTION": "test",
         "SAP_REPLY_TO_EMAIL": "test",
         "SAP_FINAL_RECIPIENT_EMAIL": "test",
         "SAP_REVIEW_RECIPIENT_EMAIL": "test",
-        "SENTRY_DSN": "test",
+        "SENTRY_DSN": None,
```
Comment (ehanson8) [inline tests/conftest.py:17]: Quoting Helen, since I did this too, this should be `"None"` since that's what it will be when read from the Dev1 env.

File: Makefile [hunk 9]
```diff
@@ -35,6 +35,9 @@ mypy:
 pylama:
 	pipenv run pylama --options setup.cfg
 
+pydocstyle:
```
Comment (ehanson8) [inline Makefile:46]: I'll PR the python-template repos soon with this change and the `isort` update

**Review by ehanson8 — COMMENTED** [review]
Looks good to me but we can see if Helen has any other feedback

File: tests/conftest.py (hunk 8 above)
Comment (adamshire123) [reply tests/conftest.py:17]: I can't tell which value you are referring to. I thought these would supercede what's in the dev1 env... Isn't that the reason for making the fixture?

Comment (ehanson8) [reply tests/conftest.py:17]: `SENTRY_DSN`, it's just a matter of making the fixture more like the real thing

Comment (hakbailey) [reply tests/conftest.py:17]: It doesn't really matter what it is here, we just need to have a test for both cases (`None` and `"None"`).

File: Makefile [hunk 10]
```diff
@@ -35,6 +35,9 @@ mypy:
...
+	pipenv run pydocstyle
```
Comment (hakbailey) [inline Makefile]: We probably only want to run pydocstyle on the sapinvoices package (not the tests), so this would be `pipenv run pydocstyle sapinvoices`.

File: sapinvoices/cli.py [hunk 11]
```diff
@@ -1,28 +1,88 @@
...
+    logger.info(configure_logger(root_logger, log_level))
     logger.info(configure_sentry())
-    logger.info("Running process")
-
-    # Do things here!
 
-    elapsed_time = perf_counter() - start_time
     logger.info(
-        "Total time to complete process: %s", str(timedelta(seconds=elapsed_time))
+        "Patronload config settings loaded for environment: %s",
```
Comment (hakbailey) [inline sapinvoices/cli.py]: Copy/paste typo?

File: sapinvoices/config.py [hunk 12]
```diff
@@ -31,3 +46,24 @@ def configure_sentry() -> str:
         sentry_sdk.init(sentry_dsn, environment=env)
         return f"Sentry DSN found, exceptions will be sent to Sentry with env={env}"
     return "No Sentry DSN found, exceptions will not be sent to Sentry"
+
+
+def load_config_values() -> dict:
+    settings = {}
+    for variable in CONFIG:
+        settings[variable] = os.getenv(variable)
+    # all settings are required in prod and stage
```
Comment (hakbailey) [inline sapinvoices/config.py]: Since most of the settings will be required in all envs, I'm not sure there's a particular reason not to just require them everywhere.

**Review by hakbailey — COMMENTED** [review]
This is ok as is but I think config could be simplified a bit (see inline comments).

**Review by hakbailey — APPROVED** [review]
👍
